### PR TITLE
the bonnet is up and we're on bricks. Fix it.

### PR DESCRIPTION
**Pull request by mankyKitty**
This is a fixathon, hacking is for the students. :smiley: 

* Update the `nix-shell` environment to more up to date definition, and include `sqlite` as it wasn't there before.
* Replace the use of `aeson` with `waargonaut`. **This is up for discussion as waargonaut is a new QFPL library and aeson is far more prevalent in the wild.**
* Improve the comments and explanations for various functions throughout the course as student feedback indicated that it was not always clear as to what was intended. In those situations where the types did not provide much guidance.
* Add another exercise involving generalising their existing `AppM` to allow for changing the error type. This allows for a more interesting challenge in the `prepareAppReqs` function initially. As well as allowing for an introduction to using the `ExceptT` transformer as a localised substitute when it is no longer possible to use `AppM`.
* Delete the application from Level03 and replace it with a more involved testing exercises where the students are actually testing their own code.
* Delete all duplicated testing. All testing will take place in the one module and students are expected to update this with their latest `Application` to check their work. As the initialisation of the `Application` becomes more involved in the latter stages, this will provide additional challenges.

#### TODO
- [x] Add documentation around using the `:memory:` file path when testing using the database.
- [x] Add initial Level08 document.

### CODE REVIEW

File: applied-fp-course.cabal [hunk 1]
```diff
@@ -118,13 +131,17 @@ library
                      , bytestring == 0.10.*
                      , text == 1.2.*
                      , optparse-applicative >= 0.13 && < 0.15
-                     , aeson == 1.*
                      , mtl == 2.2.*
                      , time >= 1.4 && < 1.10
+                     , old-locale
```
Comment (endgame) [inline applied-fp-course.cabal]: cabal gen-bounds?

File: shell.nix [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+  course = import ./. { inherit nixpkgs compiler; };
+
+  # Override the basic derivation so we can have a more fully feature
+  # environment for hacking on the course material
+  courseDevEnv = (pkgs.haskell.lib.addBuildTools course
+    [ # Include the SQLite Database application
+      nixpkgs.sqlite
+
+      # Optionally include the 'ghcid' auto reloading tool
+      # nixpkgs.haskellPackages.ghcid
```
Comment (endgame) [inline shell.nix]: This is small enough to unconditionally include?

File: src/Level02/Core.hs [hunk 3]
```diff
@@ -54,8 +54,7 @@ resp400
 resp400 =
   error "resp400 not implemented"
 
--- These next few functions will take raw request information and construct one
--- of our types.
+-- | These next few functions will take raw request information and construct one of our types.
```
Comment (endgame) [inline src/Level02/Core.hs]: This seems wrong. This attaches the text to one specific function. There are several of these comment blocks.

File: src/Level04/Types/Topic.hs [hunk 4]
```diff
@@ -1,17 +1,20 @@
...
   , getTopic
+  , encodeTopic
   ) where
 
-import           Level04.Types.Error (Error (EmptyTopic), nonEmptyText)
+import           Waargonaut.Encode   (Encoder)
+import qualified Waargonaut.Encode   as E
 
-import           Data.Aeson          (ToJSON)
+import           Data.Functor.Contravariant (contramap)
```
Comment (endgame) [inline src/Level04/Types/Topic.hs]: Realign imports?

File: src/Level06/AppM.hs [hunk 5]
```diff
@@ -6,90 +6,78 @@ module Level06.AppM where
...
+-- define this type alias to make the error type variable concrete as 'Error'.
 --
--- foo :: IO (Either Error Value)
--- foo = do
---   aE <- mightFail
---   either (pure . Left) needsAButMightFail aE
---   where
---     mightFail :: IO (Either Error Int)
---     alsoMightFail :: Int -> IO (Either Error Value)
+type AppM = AppM' Error
```
Comment (endgame) [inline src/Level06/AppM.hs]: Should `AppM'` be `AppM` and `AppM` be `App`?

File: src/Level06/Conf/File.hs [hunk 6]
```diff
@@ -1,54 +1,55 @@
...
 -- $setup
 -- >>> :set -XOverloadedStrings
 
--- | File Parsing
-
--- We're trying to avoid complications when selecting a configuration file
+-- | We're trying to avoid complications when selecting a configuration file
 -- package from Hackage. We'll use an encoding that you are probably familiar
 -- with, for better or worse, and write a small parser to pull out the bits we
 -- need. The package we're using is the ``aeson`` package to parse some JSON and
```
Comment (endgame) [inline src/Level06/Conf/File.hs]: Nope

File: src/Level04/README.md [hunk 7]
```diff
@@ -16,22 +18,42 @@ For reference, the packages we will use to talk to our database are:
...
 
 NB: The PostgreSQL example module is in ``src/Level04/DB/PostgreSQL.hs``.
 
 # Useful Typeclasses
 
+## [Contravariant](http://hackage.haskell.org/package/contravariant/docs/Data-Functor-Contravariant.html)
+
 ## [Traversable](https://hackage.haskell.org/package/base/docs/Data-Traversable.html)
 
 This typeclass provides a function called ``traverse``, which is for
```
[the remark this answers is not in the document]
Comment (mankyKitty) [reply src/Level04/README.md]: I've added a spiel that is partly a copy of the `Predicate` explanation from the haddocks, which I link to.

File: src/Level06/AppM.hs (hunk 5 above)
Comment (mankyKitty) [reply src/Level06/AppM.hs]: That's a nicer naming scheme, I've updated Level06/07 to match.

**Comment by mankyKitty** [comment]
Part of the concern that was raised about using waargonaut over aeson is that the moment you leave the course, you're going to encounter an ecosystem of packages that (almost) all use aeson.

It's a question of how well this choice aligns with the purpose of the course. Does it benefit the students or are we just using the course to push our own libraries. That's an extreme wording, but it gets the point across.

Ideally, the students will gain sufficiently general Haskell skills that they can establish how things work and use the tools they want. So our choice of aeson/waargonaut shouldn't be an issue.

My anec-data thus far is that the students enjoy these sorts of discussions and discussing pros and cons of approaches is something they want to get out of the course. So I plan to keep waargonaut in there for now, but explicitly raise it with the next class to get their opinion.

**Comment by gwils** [comment]
When I try to build this, I get errors about Level03/Core and Level03/Types listed in the exposed-modules section of the cabal file.

**Comment by mankyKitty** [comment]
Hrmm, that should be fixed already.

**Comment by gwils** [comment]
Ah yes, the trick is to git pull.

**Pull request merged**